posydon/popsyn/sample_from_file.py

```diff
@@ -0,0 +1,168 @@
+import numpy as np
+import pandas as pd
+import warnings
+from posydon.utils.posydonerror import FileError
+from posydon.popsyn.independent_sample import (generate_orbital_periods,
+                                               generate_orbital_separations,
+                                               generate_eccentricities,
+                                               generate_primary_masses,
+                                               generate_secondary_masses)
+
+PRIMARY_MASS_NAMES = ['primary_mass', 'Primary_Mass', 'PRIMARY_MASS',\
+                      'mass_1', 'Mass_1', 'MASS_1', 'm_1', 'M_1', 'm1', 'M1']
+SECONDARY_MASS_NAMES = ['secondary_mass', 'Secondary_Mass', 'SECONDARY_MASS',\
+                        'mass_2', 'Mass_2', 'MASS_2', 'm_2', 'M_2', 'm2', 'M2']
+PERIOD_NAMES = ['orbital_period', 'Orbital_Period', 'ORBITAL_PERIOD',\
+                'period', 'Period', 'PERIOD', 'P_orb', 'porb', 'P']
+SEPARATION_NAMES = ['orbital_separation', 'Orbital_Separation',\
+                    'ORBITAL_SEPARATION', 'separation', 'Separation',\
+                    'SEPARATION', 'semi-major_axis', 'Semi-Major_Axis',\
+                    'SEMI-MAJOR_AXIS', 'semi_major_axis', 'Semi_Major_Axis',\
+                    'SEMI_MAJOR_AXIS', 'a']
+ECCENTRICITY_NAMES = ['eccentricity', 'Eccentricity', 'ECCENTRICITY',\
+                      'ecc', 'Ecc', 'ECC', 'e']
+
+def infer_key(available_keys=[], allowed_keys=[]):
+    """Infer key from list of allowed keys.
+
+    Parameters
+    ----------
+    available_keys : iterable object of str, e.g. list of str
+        Collection of available keys.
+    allowed_keys : iterable object of str, e.g. list of str
+        Collection of allowed keys.
+
+    Returns
+    -------
+    key : str
+        The first matched key.
+    """
+
+    for k in allowed_keys:
+        if k in available_keys:
+            return k
+
+    return ''
+
+def get_samples_from_file(orbital_scheme='', **kwargs):
+    """Read a population of binaries at ZAMS from a file.
+
+    Parameters
+    ----------
+    orbital_scheme : str
+        Scheme to get the orbit: 'separation', 'period'
+    **kwargs : dictionary
+        kwargs from BinaryPopulation class, which should contain
+        read_samples_from_file
+
+    Returns
+    -------
+    orbital_scheme_set : ndarray of floats
+        orbital separations/periods depending on the scheme
+    eccentricity_set : ndarray of floats
+        eccentricities
+    m1_set : ndarray of floats
+        primary masses
+    m2_set : ndarray of floats
+        secondary masses
+
+    """
+
+    if 'read_samples_from_file' in kwargs:
+        filename = kwargs['read_samples_from_file']
+    else:
+        raise KeyError("In get_samples_from_file no 'read_samples_from_file'"
+                       " in kwargs.")
+    # Check and read file into dataframe
+    if os.path.isfile(filename):
+        df = pd.read_csv(filename)
+    else:
+        raise FileError(f'{filename} not found!')
```

Comment:
Do we want this to be a POSYDON error? (If the file does not exist should it be automatically thrown and not suppressed with other POSYDON errors?)

Reply:
Hm, I had the impression we'd like to capture all kinds of errors as much as possible by ourselves.
I may have a different view here than others, but for me an error is a critical issue, which requires the code the stop/not continue there, otherwise it would be a warning. Hence, errors shouldn't get suppressed unless one knows that it won't hurt to let the code crash, but usually it would be better to instead not execute the code part if it isn't needed to run properly through.

Reply:
Right, so I think with the POSYDON Errors, we are currently using them to capture errors in the code that are known but not fixable, e.g. a binary could not be matched, or there are limits with numerical methods. Any other errors that occur are not categorized as POSYDON Errors, just regular python errors, and they are always thrown. So, the default currently with the population runs is that we suppress POSYDON Errors and everything else is thrown automatically.

Reply:
So, for the binary populations, the POSYDON Errors indicated a binary failed, but there is nothing we can do about it for now. If we want this behavior to be different for other parts of the code, that is also possible.

Reply:
I have removed the FileError and replaced it by the in-build FileNotFoundError. Additionally, I added a list of in-build Error classes to the bottom of our POSYDONErrors, to give a complete overview of available errors there to allow us to always pick the most appropriate error.